contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,853 @@
+
+1. When the StafefulSet controller observes a mutation to a StatefulSet's 
+ `.Spec.Template` it will compare the `.Spec.TemplateGeneration` to the 
+ `.Status.TargetTemplateGeneration`.
+1. If the `.Spec.TemplateGeneration` is equivalent to the 
+`.Status.TargetTemplateGeneration`, no update has occurred. Note that, in the 
+event that both are nil, they are considered to be equivalent, and we expect 
+this to occur after an initial upgrade to a version of Kubernetes that supports 
+StatefulSet update form one that does not.
+1. If the `.Status.TemplateGeneration` field is nil, and the 
+`.Spec.TemplateGeneration` is not nil, then the StatefulSet has no revision 
+history. To initialize its revision history, the StatefulSet controller will 
+set both `.Status.TemplateGeneration` and `.Status.TargetTemplateGeneration` 
+to `.Spec.TemplateGeneration` and 
+[create a new PodTemplate](#podtemplate-creation). 
+1. If the `.Status.TemplateGeneration` is not nil, and if the 
+`.Spec.TemplateGeneration` is not equal to the `.Status.TargetTemplateGeneration`, 
+the StatefulSet controller will do the following.
+    1. The controller will 
+    [reconstruct the revision history](#history-reconsturction) of the 
+    StatefulSet.
+    1. If the revision history of the StatefulSet contains a PodTemplate 
+    whose `.PodTemplateSpec` is semantically, deeply equivalent to the 
+    StatefulSet's `.Spec.Template`, the youngest such PodTemplate will be used 
+    as the target PodTemplate. 
+    1. If no such PodTemplate exists, the StatefulSet controller will 
+    [create a new PodTemplate](#podtemplate-creation) from the StatefulSet's 
+    `.Spec.Template`, and it will use this as the target PodTemplate.
+    1. The controller will update the StatefulSet's `.Status.TargetTemplate` 
+    based on the selection made above.
+
+### PodTemplate Selection
+When the StatefulSet controller creates the Pods in a StatefulSet, it will use 
+the following criteria to select the PodTemplateSpec used to create a 
+Pod. These criteria allow the controller to continue to make progress toward 
+its target state, while respecting its guarantees and allowing for rolling 
+updates back and forward.
+
+1. If the StatefulSet's `.Spec.TemplateGeneration` is nil, then the cluster 
+has been upgraded from a version that does not support StatefulSet update to 
+a version that does. 
+    1. In this case the `.Spec.Template` is the current revision, 
+    and no Pods in the StatefulSet should be labeled with a 
+    `StatefulSetPodTemplateGeneration` label. 
+    1. The StatefulSet will initialize its revision history on the first 
+    update to its `.Spec.Template`.
+1. If the StatefulSet's `.Spec.TemplateGeneration` is equal to its 
+`.Status.TemplateGeneration`, then there is no update in progress and all 
+Pods will be created from the PodTemplate matching this revision.
+1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
```

Comment:
I am not sure we want to involve status fields that much in the deployment process. Isn't it simpler to always think in terms of `GenerationPartition`?